### PR TITLE
Hive 25737 fix cycle metrics

### PR DESCRIPTION
**Pull request by vcsomor**
This PR is about to improve the Compaction Initiator/Worker/Cleaner cycle metrics.

However the original PerformanceLogger approach is kept for the Initiator/Worker/Cleaner cycle metrics three new metric has been added.

## Initiator duration:
A daemon thread has been implemented for the Initiator, that measures the elapsed time since its start.

## Age of the oldest working compaction
The Age of Oldest Working Compaction metric has been implemented in the AcidMetricService.

## Age of Oldest active Cleaner metric
- CQ_CLEANER_START field added to the COMPACTION_QUEUE table
- COMPACTION_OLDEST_CLEANING_AGE metric has been added
- markCleaning method added to the CompactionTxnHandler
- ShowCompactResponseElement extended with the cleanerStart field

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThreadFactory.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+/**
+ * Factory for the {@link CycleUpdaterThread} class
+ */
+final class CycleUpdaterThreadFactory {
+
+  private static final long DEFAULT_UPDATE_INTERVAL_IN_MILLISECONDS = 1_000L;
```
Comment (lcspinter) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThreadFactory.java:25]: Do we want to make this config costumisable?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java [hunk 2]
```diff
@@ -558,4 +563,11 @@ private String getInitiatorId(long threadId) {
     name.append(threadId);
     return name.toString();
   }
+
+  private void stopCycleUpdaterThread() {
```
Comment (lcspinter) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Initiator.java:567]: ExecutorService.shutdown()

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnStore.java [hunk 3]
```diff
@@ -25,7 +25,6 @@
 import org.apache.hadoop.hive.common.ValidTxnList;
 import org.apache.hadoop.hive.common.classification.RetrySemantics;
 import org.apache.hadoop.hive.metastore.api.*;
```
Comment (lcspinter) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnStore.java:27]: I know it's not part of this change, but could you please replace this wildcard import?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThreadFactory.java (hunk 1 above)
Comment (vcsomor) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThreadFactory.java:25]: I was thinking on this while implementing...  I can do that

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThread.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.hadoop.hive.metastore.metrics.Metrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This thread simply updates a gauge metric by the given interval with the time elapsed since start
+ */
+class CycleUpdaterThread extends Thread {
```
[the remark this answers is not in the document]
Comment (vcsomor) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CycleUpdaterThread.java]: Okay... I've changed the implementation to that way

File: common/src/java/org/apache/hadoop/hive/conf/HiveConf.java [hunk 5]
```diff
@@ -3190,6 +3190,11 @@ private static void populateLlapDaemonVarsSet(Set<String> llapDaemonVarsSetLocal
         "has had a transaction done on it since the last major compaction. So decreasing this\n" +
         "value will increase the load on the NameNode."),
 
+    HIVE_COMPACTOR_INITIATOR_DURATION_UPDATE_INTERVAL("hive.compactor.initiator.duration.update.interval", "1s",
+        new TimeValidator(TimeUnit.SECONDS),
+        "Time in seconds that drives the update interval of compaction_initiator_duration metric.\n" +
+            "Smaller value results in a fine grained metric update.\n"),
```
Comment (lcspinter) [inline common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:3196]: Do you mind adding a sentence here, what will happen if this value is negative? Please mention as well, that `hive.compactor.initiator.on` must be set to `true` to make this work.

**Review by lcspinter — APPROVED** [review]
🚢 🇮🇹  (pending test)

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 6]
```diff
@@ -494,6 +494,18 @@ public static ConfVars getMetaConf(String name) {
...
+        "metastore.compactor.long.running.initiator.threshold.warning",
+        "hive.compactor.long.running.initiator.threshold.warning",
+        6, TimeUnit.HOURS,
+        "Initiator duration after which a warning will be logged. " +
+            "Default time unit is: hours"),
+    COMPACTOR_LONG_RUNNING_INITIATOR_THRESHOLD_ERROR(
+        "metastore.compactor.long.running.initiator.threshold.error",
+        "hive.compactor.long.running.initiator.threshold.error",
+        12, TimeUnit.HOURS,
+        "Initiator duration after which an wrror will be logged. " +
```
Comment (klcopp) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:507]: nit: spelling

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 7]
```diff
@@ -494,6 +494,18 @@ public static ConfVars getMetaConf(String name) {
         12, TimeUnit.HOURS,
         "Age of oldest initiated compaction in the compaction queue after which an error will be logged. " +
             "Default time unit is: hours"),
+    COMPACTOR_LONG_RUNNING_INITIATOR_THRESHOLD_WARNING(
+        "metastore.compactor.long.running.initiator.threshold.warning",
+        "hive.compactor.long.running.initiator.threshold.warning",
+        6, TimeUnit.HOURS,
+        "Initiator duration after which a warning will be logged. " +
```
Comment (klcopp) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:501]: I would say, "Initiator cycle duration" here just to be clear that it's the whole cycle we're talking about

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 8]
```diff
@@ -379,6 +379,102 @@ public void markCompacted(CompactionInfo info) throws MetaException {
     }
   }
 
+  /**
+   * Mark the cleaning start time for a particular compaction
+   *
+   * @param info info on the compaction entry
+   */
+  @Override
+  @RetrySemantics.CannotRetry
```
Comment (klcopp) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:388]: I'm not 100% on these, but shouldn't this have the @RetrySemantics.ReadOnly annotation?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/metrics/MetricsConstants.java [hunk 9]
```diff
@@ -22,9 +22,13 @@
   public static final String API_PREFIX = "api_";
   public static final String COMPACTION_STATUS_PREFIX = "compaction_num_";
   public static final String COMPACTION_OLDEST_ENQUEUE_AGE = "compaction_oldest_enqueue_age_in_sec";
+  public static final String COMPACTION_OLDEST_WORKING_AGE = "compaction_oldest_working_age_in_sec";
+  public static final String COMPACTION_OLDEST_CLEANING_AGE = "compaction_oldest_cleaning_age_in_sec";
   public static final String COMPACTION_INITIATOR_CYCLE = "compaction_initiator_cycle";
+  public static final String COMPACTION_INITIATOR_DURATION = "compaction_initiator_duration";
```
Comment (klcopp) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/metrics/MetricsConstants.java:28]: I would change this to compaction_initiator_cycle_duration in order to be very clear; similar for compaction_cleaner_duration

**Review by klcopp — COMMENTED** [review]
Looks good, a couple comments/questions

**Review by klcopp — APPROVED** [review]

**Review by klcopp — APPROVED** [review]

**Pull request merged**